Add a zngio scanner

nwt opened the pull request:
This implements steps 1 through 3 of #681.

* Change the ZNG specification to allow only value messages in a
  compressed message block.

* Add scanner.ScannerAble and scanner.Scanner interfaces.

* Modify zngio.Reader to implement scanner.ScannerAble with efficient
  handling for compressed value message blocks.

* Add scanner.NewMultiScanner with knowledge of scanner.ScannerAble.

* Change driver.Compile to accept a []zbuf.Reader parameter instead of a
  single zbuf.Reader and to use scanner.NewMultiScanner.

nwt commented:
System tests are broken, but I wanted to get this out before fixing them.

philrz commented:
I read through the spec changes FWIW and they look good to me. 👍

mattnibs commented on scanner/scanner.go:
```diff
@@ -4,16 +4,53 @@ import (
+}
+
+// Accumulate updates its receiver by adding to it the values in ss.
+func (s *ScannerStats) Accumulate(ss *ScannerStats) {
```
Nice!

mattnibs commented on driver/compile.go:
```diff
@@ -28,7 +27,7 @@ type Config struct {
 // proc.MuxOutput that which brings together all of the flowgraphs
 // tails, and is ready to be Pull()'d from.
-func Compile(ctx context.Context, program ast.Proc, zctx *resolver.Context, reader zbuf.Reader, cfg Config) (*MuxOutput, error) {
+func Compile(ctx context.Context, program ast.Proc, zctx *resolver.Context, readers []zbuf.Reader, cfg Config) (*MuxOutput, error) {
```
I'd prefer to do this:
```suggestion
func Compile(ctx context.Context, program ast.Proc, zctx *resolver.Context, cfg Config, readers ...zbuf.Reader) (*MuxOutput, error) {
```

nwt replied:
I've removed this change to the signature of driver.Compile.

mattnibs commented on scanner/multiscanner.go:
```diff
@@ -0,0 +1,85 @@
+// read sequentially, and when all records are consumed, Pull will return a nil
+// batch and nil error. If any reader returns a non-nil error, Pull will return
+// that error.
+func NewMultiScanner(ctx context.Context, readers []zbuf.Reader, filterExpr ast.BooleanExpr, span nano.Span) (Scanner, error) {
```
We don't have to do this now but at some point it we should consider adopting the Options pattern here:
```suggestion
func NewMultiScanner(ctx context.Context, readers []zbuf.Reader, opts ...Options) (Scanner, error) {
```

nwt replied:
I've removed this function.

nwt commented:
The driver.Compile changes here (particularly use of scanner.NewMultiScanner) conflict with work @alfred-landrum is doing for #1040 and will probably be discarded. In light of that, I've converted this PR to a draft. I'll mark it ready for review after I've updated it to align with the #1040 work.

mattnibs commented on zbuf/combiner.go:
```diff
@@ -32,27 +32,33 @@ func CmpTimeReverse(a, b *zng.Record) bool {
-	return &Combiner{
+func NewCombiner(readers []Reader, cmp RecordCmpFn) ReadCloser {
+	if len(readers) == 1 {
+		// XXX
```
Seems like a reasonable check, can this comment be removed?

nwt replied:
It certainly can. Done in 79d7a50f8d77978dcfb75a3beb8f1a7383403385.

nwt commented on zng/docs/spec.md:
```diff
@@ -351,27 +351,26 @@ previously defined type, the appropriate typedef control code must
-block comprises a compressed sequence of control and value messages.
-The sequence can include any message except for end-of-stream markers
-and compressed message blocks.
+Following a header byte of 0x86 is a compressed value message block.
+Such a block comprises a compressed sequence of value messages.  The
+sequence must not include control messages.
```
I made this ZNG specification change in anticipation of adding Boyer-Moore string search inside zio/zngio.zngScanner.scanUncompressed. Excluding control messages like this means we can immediately discard a block if the search fails, without looking for control messages.

It accounts for the the changes in ZNG outputs and the zng-output-check failure.

alfred-landrum commented on zio/zngio/reader.go:
```diff
@@ -126,29 +141,24 @@ again:
+				return 0, nil, err
 			}
-			err = r.readCompressed()
+			return -zng.CtrlCompressed, r.uncompressed.Bytes(), nil
```
This is being exposed like this in preparation for the boyer-moore work, right? (instead of just reading the compressed data & then falling out to the `goto again`, like the `CtrlEOS` handling)

alfred-landrum replied:
never mind, I read down to `zngScanner.Pull`
